keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1256 @@
+changing the invalid fields during an update.
+
+##### Unit tests
+
+<!--
+In principle every added code should have complete unit test coverage, so providing
+the exact set of tests will not bring additional value.
+However, if complete unit test coverage is not possible, explain the reason of it
+together with explanation why this is acceptable.
+-->
+
+<!--
+Additionally, for Alpha try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+The data can be easily read from:
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `apiextensions-apiserver/pkg/apiserver/schema/cel`: `<date>` - `<test coverage>`
+- `apiextensions-apiserver/pkg/apiserver/schema/listtype`: `<date>` - `<test coverage>`
+- `apiextensions-apiserver/pkg/apiserver/schema/objectmeta`: `<date>` - `<test coverage>`
+
+##### Integration tests
+
+<!--
+Integration tests are contained in k8s.io/kubernetes/test/integration.
+Integration tests allow control of the configuration parameters used to start the binaries under test.
+This is different from e2e tests which do not allow configuration of parameters.
+Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
+-->
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+1. Show normal creation path is unaffected
+2. Tests that ratchet different classes of validations:
+    - JSON-Schema fields (minItems, dependencies, etc.)
+    - CRD-Validation Rules
+    - ListSets/Maps
+3. Tests that certain classes of valdiations are not ratcheted:
+    - Nested Validations (not, oneOf, allOf, anyOf)
+    - Transition Rules
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+It would be nice to have a e2e test that tests ratcheting over a k8s upgrade
+which tightens a validation.
+
+- <test>: <link to test coverage>
+
+### Graduation Criteria
+
+
+#### Alpha
+- Feature Flag added
+- Integration Tests implemented and enabled
+- Schema-Based Ratcheting Implemented
+- List-Type Ratcheting Implemented
+- ObjectMeta/TypeMeta object-level ratcheting
```

Comment:
nit: don't think all those three cases are a prereq for an alpha technically.

Background: completeness is not even necessary for GA.

If review time or issues become a problem, we can start with a partial implementation (compare your conversativeness argument above).

Reply:
Moved these requirements to beta and added beta requirement of expanded discussion of types of ratcheting for each type of validation rule

Reply:
Alpha now has requirement of an incomplete-but-correct version of ratcheting implemented